INF-0004 Specify diagnostics for resource binding, specifically on resource types and binding prefixes. (#230)
microsoft/hlsl-specs: proposals/infra/INF-0004-register-types-and-diagnostics.md

class that are "sampleable". These resources are typically
prefixed with "Texture" in actual HLSL source. So, there will
be a special check to see whether the candidate resource is 
prefixed with "Texture", and if so, the "s" register type will
also be permitted for that resource.

If the given candidate resource is a user-defined type (UDT), then further
analysis is necessary. The first step is to gather all registers that
are being bound to this declaration, and collect the register types
that are being specified. The UDT must have at least one valid resource
that can be bound to the provided register type. If not, an error must be
emitted stating that "No resource contained in struct '%0' can be bound
to register type '%1'". There are no issues if a UDT has more resources
than there are register binding statements, the resources will be bound to
the next available space automatically, and so compilation can succeed.
Additionally, a singular resource can be bound multiple times to different
spaces, so multiple register types are allowed for a single resource.
Below is a table that represents the different behaviors that could arise
given some examples of different UDT's:

> | Code | Diagnostic |

Reviewer: github doesn't seem to be formatting this as a table any more.

I think that this would be much easier to write and read if it were just flat.  Something like:

```
struct Eg1 {
  float f;
  Buffer<float> Buf;
  RWBuffer<float> RWBuf;
};
Eg1 e1 : register(t0) : register(s0); 
// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to s0
```
...
```
struct Eg5 {
  float f;
};

Eg5 e5 : register(t0);
// DefautError warning. "UDF resource 'Eg5' does not contain an applicable resource type for register type 't'.
```

Reviewer: Does it look like this for you?

![image](https://github.com/microsoft/hlsl-specs/assets/8118402/b04ad489-2bff-430e-a0a2-35da97600da7)

Note that all the indentation is broken, and the code is all scrunched up on the left so that it wraps and makes it hard to read.

| struct Foo {<br>  float f;<br>  Buffer<float> Buf;<br>}; <br> Foo f : register(t0) : register(s0); | None. Success because Buf gets bound to t0. Buf will also be bound to s0.|
| struct Foo {<br>  struct Bar {<br>    RWBuffer<int> a;<br>    };<br>    Bar b;<br>};<br> Foo f : register(t0) | None. Success because Bar, the struct within Foo, has a valid resource that can be bound to t0. |
| struct Foo {<br>  float f;<br>}; <br> Foo f : register(t0) | DefaultError warning. "UDT resource 'Foo' does not contain an applicable resource type for register type 't'"|
| struct Foo {<br>  struct Bar {<br>      float f;<br>    }<br>    Bar b;<br>};<br> Foo f : register(t0) | DefaultError warning. "UDT resource 'Foo' does not contain an applicable resource type for register type 't'"|

Finally, if the candidate type is not a valid resource type or not a UDT,
the final case will be entered. Types that can be immediately determined
to not be a resource type or not contain a resource type (types like booleans,

> int, float, float4, etc) are known as "non-intangible". This is because

Reviewer: Ah, ok.  I thought maybe "non-intangible" was a mistake.  To avoid the double-negative, how about calling them "tangible"?

## Motivation

There are several cases in DXC in which diagnostics that are related to
invalid register types are out of date or invalid. For example, in the case of:

`float b : register(u4);`
an error will be emitted recommending the use of the 'b, c, or i' register
type, however these register types are no longer in support. Additionally,
it is possible the user is unaware that this resource won't actually be
used, but the compiler doesn't communicate that to the user.
It would be great for any HLSL developer to immediately have answers to 
the following questions:
For any valid resource type that needs to be bound, what are the set of
valid register types for that resource, and what are all the possible 
diagnostics that can be emitted if the resource is given a register 
type that isn't in the set of valid register types?
The design below aims to specify the answer to these questions.

## Proposed solution
Firstly, the most common case is when a resource type that is not a 
user-defined type is bound as a resource. Any resource will have an 
associated ResourceAttr attribute, from which we can determine the 
resource class and resource kind. These two inputs are sufficient to

> determine the expected register type, and any other register type

Author: Right, fixed

will result in an error. The recommended register type will be
suggested in the diagnostic. The table below specifies what register 
type will be expected for any resource type that falls under the
specified resource class. 

| Resource Class | Register Type | Diagnostic |
|-|-|-|
| Sampler | s | "Object type '%0' with resource class "Sampler" expects register type 's', but register type '%1' was given." |
| SRV | t | "Object type '%0' with resource class "SRV" expects register type 't', but register type '%1' was given." |
| UAV | u | "Object type '%0' with resource class "UAV" expects register type 'u', but register type '%1' was given." |
| CBuffer | b |  "Object type '%0' with resource class "CBuffer" expects register type 'b', but register type '%1' was given." |

> If the given candidate resource is a user-defined type (UDT), then further

Author: Fixed up the language, hopefully this is better?

> If the given candidate resource is a user-defined type (UDT), then further

Reviewer 2: Still says "candidate resource", when it should say something like "candidate variable" instead.

analysis is necessary. The first step is to gather all registers that
are being bound to this declaration, and collect the register types
that are being specified. The UDT must have at least one valid resource
that can be bound to the provided register type(s). If not, an error must be
emitted stating that "No resource contained in struct '%0' can be bound
to register type '%1'". There are no issues if a UDT has more resources
than there are register binding statements, the resources will be bound to
the next available space automatically, and so compilation can succeed.
Additionally, a singular resource can be bound multiple times to different

> spaces, so multiple register types are allowed for a single resource.

Author: Thanks for the clarification, I was mistaken.

struct Eg2 {
  float f;
  Buffer<float> Buf;
  RWBuffer<float> RWBuf;
  RWBuffer<float> RWBuf2;
  };
Eg2 e2 : register(t0) : register(s0); 
// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to s0. RWBuf2 gets assigned to s1 even though there is no explicit binding for s1.

struct Eg3 {
  float f;
  Buffer<float> Buf;
  }; 
Eg3 e3 : register(t0) : register(s0);
// Valid: Buf gets bound to t0. Buf will also be bound to s0.

struct Eg4 {
  struct Bar {
    RWBuffer<int> a;
    };
    Bar b;
};
Eg4 e4 : register(t0) 
// Valid: Bar, the struct within Eg4, has a valid resource that can be bound to t0. 

struct Eg5 {
  float f;
}; 
Eg5 e5 : register(t0) 
// DefaultError warning: "UDT resource 'Eg5' does not contain an applicable resource type for register type 't'"

struct Eg6 {
  struct Bar {
      float f;
    }
    Bar b;
};
Eg6 e6 : register(t0) 
// DefaultError warning: "UDT resource 'Eg6' does not contain an applicable resource type for register type 't'"|

```

Finally, if the candidate type is not a valid resource type or not a UDT, the
final case will be entered. Types that are or contain a resource are known as

> "intangible". In this case, we are dealing with types that cannot be intangible.

Author: I do think `intangible` is the proper word to use here. Do you have an alternate phrase to use to describe types that don't have the resource attribute or do not contain any types that have the resource attribute? Ideally, the spec can refer to this concept with a single word, rather than needing to repeat that whole phrase several times.

> "intangible". In this case, we are dealing with types that cannot be intangible.

Reviewer 2: You're right that we could use some compact language to refer to these types.  These basic types are "tangible" (or not "intangible), but that also applies to UDTs that don't contain intangible types.  I'd like to think of these as "basic" types in HLSL, but might conflict with Clang, since "basic" types in Clang probably don't include our vector and matrix types.

Types that can be immediately determined to not be intangible (that is, types that
cannot be a resource type or cannot contain a resource type) are types like booleans,
int, float, float4, etc. If the resource type is among any numerics or a type that
cannot be an intangible type, there is only one exception where it can be treated
as a resource. The only valid register type for such resource types is 'b', which

> adheres to legacy behavior. All other register types applied to such a type will 

Reviewer: I'm not sure what "adheres to legacy behavior" means (in particular "adheres").

I wonder if having a separate section/paragraph to explain the legacy behavior might help me follow this.

What does it mean that it is legacy? What are we being compatible with here? Why do we think it is important to continue to support this?

Author: I've addressed this in the spec, hopefully it's clear now. legacy means DX9 in this spec. I expounded a bit to explain the behavior. It is important to continue to support this because developers actively use the 'c' register overload to offset variables in the $Globals constant buffer.

emit an error diagnostic that "the given type '%0' cannot be bound as a resource".
If the register type is 'b', a warning will be emitted instead, that will be
treated as an error by default: "register type 'b' used for resource type '%0', 
which cannot be used as a resource". Below are some examples:

| Code | Diagnostic |
|-|-|
| `float f : register(t0)` | "error: 'float' is an invalid resource type for register type 't'" |

> | `float f : register(b0)` | DefaultError "warning: register type 'b' used for resource type 'float', which cannot be used as a resource. Disable with --Wno-disallow-legacy-binding-rules" |

Reviewer: The `b0` diagnostic doesn't appear in the flowchart.  In the flowchart there's:

* warning: register type 'b' used for resource type '<T>', which cannot be used as a resource.
* DefaultError warning: Using register type '<p>' is deprecated, resource won't be used. Disable with --Wno-disallow-legacy-binding-rules

Neither of which match the one shown in this table.

Author: Thanks, added the disable clause

## Detailed design

All the compiler has to work with is a Decl object. From the Decl object, 
the ResourceAttr attribute can be fetched, with which the resource 
class can be determined. Once this is known, we then check if the resource

> type is a user-defined type or if it is a standard resource type. If the 

Author: Extraordinarily helpful, especially the default_globals part! Thanks for this contribution, I've baked it into the spec and the flowcharts.

to be validated to see whether it can properly be assigned to the given register 
type. If the UDT does not have at least one resource that matches the given register 
types, then an error will be emitted. The only remaining case is when there
is no resource attribute, which implies the type cannot be intangible. In this case,
if the type is strictly numeric, the only acceptable register type is 'b', which 
is legacy behavior. In this case, a warning will be emitted that will be treated as 
an error by default, which states that a numeric type is not usable as a resource. 
If 'c' or 'i' are given as register types, a warning will be emitted stating that
such register types are legacy behavior, and the resource won't be used. Any other 
register type will cause an error to be emitted, stating that the given resource
type cannot be bound as a resource.
Legacy behavior can be allowed with the --Wno-disallow-legacy-binding-rules
flag. When this flag is active, and legacy behavior is present, a warning will 
be emitted instead of an error.

Below is a flowchart that describes the process that is used to determine
what kind of diagnostic to emit in each case:

```mermaid

> flowchart TD

Reviewer: Now that the rules are simpler, I wonder if the flowchart is the clearest way to represent this?

Author: Tex's feedback has made me aware that this is more complex than previously thought. I think the flowchart should stay now.

cannot be an intangible type, there is only one exception where the register keyword
can legally be applied to the variable. The only valid register type for such a 
numeric variable type is 'c'. The register keyword in this statement:
`float f : register(c0)`
isn't binding a resource, rather it is specifying a constant register binding offset
within the $Globals cbuffer, which is legacy behavior from DX9. Because this specific 
register overload is only applicable if the developer wants to make modifications to the
$Globals buffer, it is not useful when the overload is not used inside the $Globals
context. That is, if this register overload appears within a `cbuffer {...}` or 
`tbuffer {...}` scope, the register number will be ignored. In this case, a warning will
be emitted that is treated as an error:
`warning: variable of type '%0' bound to register type '%1' does not contain a matching '%select{SRV|UAV|CBV|Sampler}2' resource`
All other register types applied to such a type will emit an error diagnostic that 
`"'%0' is an invalid resource type for register type '%1'"`, except for 'b' or 'i'.
If the register type is 'b' or 'i', a warning will be emitted instead, that will be 

> treated as an error by default. <br>
> Below are some examples:

Reviewer 2: Not sure why embedded `<br>` is showing up here, but it's probably better to just use a newline instead.  There are other examples of this later as well.
```suggestion
treated as an error by default.

Below are some examples:
```

| `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used. Disable with --Wno-disallow-legacy-binding-rules" |
| `float f : register(x0)` | "error: register binding type 'x' not supported for variable of type 'float'" |
| `cbuffer g_cbuffer { float f : register(c2); }` | "warning: register binding 'c' should only be used in global contexts. Disable with --Wno-disallow-legacy-binding-rules" |
| `struct Foo { RWBuffer<float> f; }; Foo foo : register(c2);`| "warning: 'c' register type should only be used for basic types. Disable with --Wno-disallow-legacy-binding-rules" |
| `RWBuffer<float> f : register(c3);`| "error: invalid register specification, expected 'u' binding" |

## Detailed design

All the compiler has to work with is a Decl object and the context the decl appears in.
From this information, the first goal is to set a specific set of flags that can fully inform
the right diagnostic to emit. The first group of flags are the decl type flags, which
are either `basic`, `resource`, `udt`, or `other`. `basic` refers to a numeric variable
type. `other` refers to a type that cannot possibly be resource, because it lacks the 

> resource attribute.<br>

Reviewer 2: `other` actually refers to other special object types (intangible types) that are not resources, and thus cannot be placed into a cbuffer or bound as a resource.

The next group of flags are the resource class flags,
which are `srv`, `uav`, `cbv`, or `sampler`.<br>
The final flag is `default_globals`, which indicates whether or not the decl appears
inside the $Globals scope.<br><br>

From the Decl object, first determine if the decl is a cbuffer or tbuffer decl. If so,
we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
respectively). Otherwise, the decl is classified according to the underlying canonical 
type.<br>

> If the decl is implicit, then check if the decl is a vector, matrix, or otherwise numeric type,

Reviewer 2: ```suggestion
If the decl is implicit, then check if the decl is a vector, matrix, or basic numeric or bool type,
```

> If the decl is implicit, then check if the decl is a vector, matrix, or otherwise numeric type,

Reviewer 2: ```suggestion
If the decl is implicit, then check if the decl is a vector or matrix,
```
What do you mean by "or otherwise numeric type"?

> and if so, we know that the decl is a `basic` type. If the decl has a resource attribute, 
> we can set the `resource` flag, and can also obtain the resource class and set the resource 
> class flag accordingly. Otherwise, we know the decl is not a resource and not basic, and so
> the `other` flag is set.<br>

Reviewer 2: If it's a vector or matrix, it's not a resource, so there's no resource handling here.
```suggestion
and if so, we know that the decl is a `basic` type. Else if the decl has a resource attribute, 
we can set the `resource` flag, and can also obtain the resource class and set the resource 
class flag accordingly. Otherwise, we know the decl is not a resource and not basic, and so
the `other` flag is set.
```

Author: Reworded it myself.

> If the decl is not implicit, then we check if it is a struct or class type, and if so, the
> decl is a UDT, so the `udt` flag is set. We then recurse through the members of the UDT and
> collect information on what types of resources are contained by the UDT (uav, cbv, srv, or
> sampler). The corresponding resouce class flags are set. If the decl is not a struct or class
> type, then it is a basic type, and so the `basic` flag is set.<br>

Reviewer 2: ```suggestion
If the decl type is not implicit, then we check if it is a struct or class type.
If so, the decl is a UDT, so the `udt` flag is set, otherwise, it's a basic type,
so we set the `basic` flag.

For a `udt`, recurse through the members of the UDT and its bases,
collecting information on what types of resources are contained by the UDT,
setting corresponding class flags (`uav`, `cbv`, `srv`, or `sampler`).
Also track whether there are any other numeric members and set a
`numeric` flag.

```

> The last step is to simply check if the decl is in the `$Globals` scope. If so, then the
> `default_globals` flag will be set.<br><br>

Reviewer 2: Technically `$Globals` isn't a "scope", so we might need to be more precise in the language here.
```suggestion
If the global decl is not inside a cbuffer or tbuffer block, set the `default_globals` flag.
```

Now enough information has been gathered to determine the right diagnostics to emit, if any.

> Firstly, if `other` is set, then the variable type is not valid for the given register type.
> So, we can emit this diagnostic: `error: '%0' is an invalid resource type for register type '%1'`

Reviewer 2: If `other` is set, the type isn't something that can be used with the 'register' annotation at all, so I think the message should reflect that.  My original suggestion was the message:
`"register binding not allowed on variable of type '%0'"`

```suggestion
Firstly, if `other` is set, then the variable type does not support `register` binding,
So, we can emit this diagnostic: `register binding not allowed on variable of type '%0'`

```

> If the `resource` flag is set and the `udt` flag is not set, then we have to check the 
> register statement, and the resource class flag, and verify that the register type matches the 
> resource class flag. If not, we emit this diagnostic: 
> `error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`

Reviewer 2: `resource` and `udt` should be mutually exclusive.  Either the type is a single resource type and `resource` is set, or it's a user defined type (that could potentially contain resources) and `udt` is set.

```suggestion
If the `resource` flag is set, check the register type against the resource class flag.
If they do not match, emit this diagnostic:
`error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`

```

If the `basic` flag is set, then we first check if `default_globals` is set. If so, then 
we check the register type. If the register type is 'i' or 'b', the deprecated warning

> that is treated as an error as shown in the examples above will be emitted. If 'c' is given,
> no errors will be emitted. If `default_globals` is not set and 'c' is given, a warning will
> be emitted that is treated as an error by default: `warning: register binding 'c' should only be used in global contexts. Disable with --Wno-disallow-legacy-binding-rules` 
> After this point, `default_globals` doesn't need to be set.

Reviewer 2: Suggested rewording, and `-Wno-disallow-legacy-binding-rules` shouldn't be part of the message, should it?  It will emit that warning group as part of the message automatically.  Also, I think `pack_offset` should be suggested, since it does the thing that would otherwise be done by the `c` binding inside cbuffer/tbuffer decls.

```suggestion
that is treated as an error as shown in the examples above will be emitted.
If the register type is 'c', it is allowed, but if `default_globals` is not set,
a warning that is treated as an error by default will be emitted:
`warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead`.
This warning is part of the warning group `disallow-legacy-binding-rules`.

```

> If 't', 'u', or 's' are given, then this diagnostic will be emitted:
> `error: '%0' is an invalid resource type for register type '%1'`. If any other register type
> is seen, this error will be emitted:
> `error: register binding type '%1' not supported for variable of type '%0'`<br><br>

Reviewer 2: This implies that '%0' is some resource type, but it's not.  My suggestion was:
`unsupported resource register binding '%select{t|u|b|s}1' on variable of type '%0'`

```suggestion
If 't', 'u', or 's' are given, then this diagnostic will be emitted:
`unsupported resource register binding '%select{t|u|b|s}1' on variable of type '%0'`.
If any other register type is seen, this error will be emitted:
`error: register binding type '%1' not supported for variable of type '%0'`

```

> Finally, in the case that `udt` is set, we first check `default_globals`. If it is set,
> then we can permit the 'c' register type if the UDT has at least one numeric member. If
> there is no numeric member, a warning that is treated by default as an error will be 
> emitted: `"warning: 'c' register type should only be used for basic types. Disable with --Wno-disallow-legacy-binding-rules"`
> Otherwise, cbuffers and tbuffers are not permitted within UDT's, and so we don't need to
> check if 'c' is given within the UDT nor emit a warning for legacy behavior that is 
> treated as an error. An error will already be emitted preventing the declaration of 
> these buffers within a UDT. After this point, `default_globals` doesn't need to be set.

Reviewer 2: This is confusing, and I don't think it's correct.  And no need to talk about something that's just invalid syntax (trying to define a cbuffer or tbuffer block inside a struct or class definition).

```suggestion
Finally, handle the `udt` case.

For register type 'c': if `default_globals` is not set, we emit the same default-error
warning as for `basic` types; otherwise, if `numeric` is not set, we emit a warning
that is treated as an error by default:
`warning: register 'c' used on type with no contents to allocate in a constant buffer`

```

For every register type that is used to bind the resources contained in the given UDT, 
we verify that the corresponding resource class flag has been set. If the
corresponding resource class flag is not set, this error will be emitted:
`error: variable of type '%0' bound to register type '%1' does not contain a matching '%select{SRV|UAV|CBV|Sampler}2' resource`
Otherwise, if any other register type is given, then we emit this error:
`error: register binding type '%0' not supported for variable of type '%1'`<br><br>

> Legacy behavior can be allowed with the --Wno-disallow-legacy-binding-rules

Reviewer 2: "legacy-binding-rules" is a bit overloaded at this point, since we actually have something like `-flegacy-resource-reservation` which could be thought of as "legacy binding rules".  I think we should change the name to something like `disallow-unused-register-bindings`.

The overall approach this diagnostic infrastructure will take
in emitting the right diagnostic can be broken down into two steps:
1. Analyze the Decl and set the appropriate flags.
2. Emit a diagnostic driven solely by which flags are set.
Some common cases will be simply described below, and then
the analysis step will be specified, with the diagnostic emission
step right afterwards.

### Simple explanations for common cases

The most common case is when a resource type that is not a 
user-defined type is bound as a resource. Any resource will have an 
associated ResourceAttr attribute, from which we can determine the 
resource class, which is sufficient to determine the expected register
type. Any other valid register type (among 't', 'u', 'b', or 's') will 
result in `err_hlsl_mismatching_register_type_and_resource_type`.
The recommended register type will be suggested in the diagnostic.
Any other invalid register type will emit `err_hlsl_mismatching_unsupported_register_type_and_variable_type`.
The table below specifies what register type will be expected for 
any resource type that falls under the specified resource class. 

> | Resource Class | Register Type | Diagnostic |

Reviewer: This table doesn't render correctly in github's preview.

register overload is only applicable if the developer wants to make modifications to the
$Globals buffer, it is not useful when the overload is not used inside the $Globals
context. That is, if this register overload appears within a `cbuffer {...}` or 
`tbuffer {...}` scope, the register number will be ignored. In this case, 
`warn_hlsl_register_type_c_not_in_global_scope` will be emitted, and will be treated as an error by default.
All other register types applied to such a type will emit `err_hlsl_unsupported_register_prefix"`, except for 'b' or 'i'.
If the register type is 'b' or 'i', then `warn_hlsl_deprecated_register_type_b` or `warn_hlsl_deprecated_register_type_i` 
will be emitted respectively, and treated as errors by default.

Below are some examples:

| Code | Diagnostic |
|-|-|
| `float f : register(t0)` | "error: unsupported resource register binding 't' on variable of type 'float'" |
| `float f : register(c0)` | Valid, no errors
| `float f : register(b0)` | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." |

> | `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used." |

Reviewer: I've been digging and found some references to these things in the fxc source. The only reference I've found on learn.microsoft.com is in the [ASM reference](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx9-graphics-reference-asm-ps-registers-ps-3-0).

> | `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used." |

Author: This looks like a valid use of the 'i' register type in FXC:
https://shader-playground.timjones.io/45ddf033843147d57a13312f2223dd8a

| `cbuffer g_cbuffer { float f : register(c2); }` | "warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead" |
| `RWBuffer<float> f : register(c3);`| "error: UAV type 'RWBuffer<float>' requires register type 'u', but register type 'c' was used." |

## Detailed design

### Analysis 

All the compiler has to work with is a Decl object and the context the decl appears in.
From this information, the first goal is to set a specific set of flags that can fully 
inform the compiler of the right diagnostic to emit. The first group of flags are the 
decl type flags, which are either `basic`, `resource`, `udt`, or `other`. `basic` refers 
to a numeric variable type. `other` refers to built-in HLSL object types (intangible types) 
that are not resources, and thus cannot be placed into a cbuffer or bound as a resource.

These flags are all mutually exclusive, and are represented in the code with an enum.
The `udt` flag has an associated flag, `contains_numeric`, which will be set if the
UDT contains a numeric type.

The next group of flags are the resource class flags,

> which are `srv`, `uav`, `cbv`, or `sampler`.

Reviewer 3: I see references to 'other' flag below.  It doesn't appear in this list of supported flags..  Should it be here?  With a definition of what it represents?

Author: `other` is in the "decl type" group of flags, and there's a short explanation in the first paragraph describing what it represents.

The final flag is `default_globals`, which indicates whether or not the decl appears
inside the $Globals scope.

> From the Decl object, first determine if the decl is implicit and intangible. If so,

Reviewer: > first determine if the decl is implicit and intangible

This is the first occurrence of the word "implicit" in this document. I don't quite get how you could a register binding on something that's only declared implicitly?

> then we determine if the decl is a cbuffer or tbuffer decl

how do we do that?

Author: I now mention the dynamic casting approach in the spec. I've also taken out the word "implicit" from the spec entirely.

then we determine if the decl is a cbuffer or tbuffer decl. If so,
we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
respectively). Otherwise, check to see if the ResourceAttr attribute is present. If
it is, the `resource` flag can be set, and the corresponding resouce class flag will
be set. Otherwise, the `other` flag is set. 

If the decl is implicit but not intangible, then check if the decl is a vector, matrix, or otherwise numeric type,
and if so, we know that the decl is a `basic` type. If the decl isn't any of those types, then 
check if the decl has a resource attribute. If so, we can set the `resource` flag, 

> and the corresponding resouce class flag will be set. Otherwise, we know the

Reviewer 3: nit: resouc = resource

`warn_hlsl_UDT_missing_basic_type`, are all within a warning group known as 
`disallow-legacy-binding-rules`. If legacy behavior is desired, a user can pass the `--Wno-disallow-legacy-binding-rules`
flag to the compiler to silence the errors. When this flag is active, and legacy behavior is present, a warning will 
be emitted instead of an error.

Here is some pseudocode summarizing the diagnostic emission process:

```
if other is set:
  emit err_hlsl_unsupported_register_type_and_variable_type

if is_member is set:
  emit err_hlsl_register_annotation_on_member

if resource is set:
  if register type does not match resource class flag:
    emit err_hlsl_mismatching_register_type_and_resource_type

if basic is set:
  if default_globals is set:
    if register type is 'b' or 'i':
      emit warn_hlsl_deprecated_register_type_b or warn_hlsl_deprecated_register_type_i respectively

  if register type is 'c'
    if default_globals is not set:
      emit warn_hlsl_register_type_c_not_in_global_scope
  else if register type is 't' 'u' or 's'
    emit err_hlsl_mismatching_register_type_and_variable_type
  else emit err_hlsl_unsupported_register_type_and_variable_type

> if udt is set:

Reviewer: I'm wondering if we can get rid of `udt` here and come up with something more general that just works on everything.

I also note that `basic` and `contains_numeric` mean _almost_ the same thing, and if we stopped differentiating between udt and non-udt then they would be the same thing.

Author: My initial reaction to this proposition is that it conflicts with the idea of not emitting any warning if a UDT doesn't have a resource type that's required by a specified register annotation. If UDT's are treated the same as standard resources, then we need to make some sort of distinction so that:
RWBuffer<int> r : register(s1); emits an error. Because if r's type was a UDT, and it didn't have a sampler, it wouldn't emit an error, but *because* it's not a UDT, it should emit an error.

  if default_globals is set:
    if contains_numeric is not set and the register type is 'c':

>       emit warn_hlsl_UDT_missing_basic_type

Author: Here's an example in DXC where a UDT that contains a numeric type gets placed into the $Globals cbuffer.
I agree that default_globals should only be set for values that end up in $Globals, and it isn't as simple as checking if it's in the global scope. I've modified the text in the spec to more accurately define when default_globals should be set, but this case would certainly be a "set default_globals" case.

```
struct Foo {
    float f;
    SamplerState s[3];
    };

Foo f : register(u5) : register(c1);
Texture2D tx : register(s1);

float4 main() : SV_TARGET {
    return tx.Sample(f.s[0], float2(0,0)) * f.f;
}
```

>       emit warn_hlsl_UDT_missing_basic_type

Reviewer: Yes, I agree that `Foo f` should have `default_globals` set.  Can you construct a case where `default_globals` is set but `contains_numeric` is not set?

>       emit warn_hlsl_UDT_missing_basic_type

Author: No, I don't think such a case exists. 
I've changed the pseudo-code to account for this.

  for each register annotation r applied to the udt decl:
    if r has register type t:
      if srv is not set:
        emit warn_hlsl_UDT_missing_resource_type_member
    else if r has register type u:
      if uav is not set:
        emit warn_hlsl_UDT_missing_resource_type_member
    else if r has register type b:
      if cbv is not set:
        emit warn_hlsl_UDT_missing_resource_type_member
    else if r has register type s:
      if sampler is not set:

>         emit warn_hlsl_UDT_missing_resource_type_member

Author: After some discussion, we agreed to only emit the warning as a warning, and it can be silenced by the developer if it's annoying.

float f1 : register (c0); // basic (ends up in $Globals constant buffer, so default_globals is set)
```

The first step is to simply check if the decl is inside a cbuffer or tbuffer
block. If it isn't, and it has a numeric type, set the `default_globals` flag,
because the value is bound to be placed in the `$Globals` buffer. If the decl is
in a class or struct, it is a member declaration, and register annotations aren't
allowed here, so we must set the `is_member` flag.

Next, determine if the decl is a `cbuffer` or `tbuffer`. These are special in that they have
their own decl class type, `HLSLBufferDecl`. To determine if the Decl is a `HLSLBufferDecl`,
the compiler dynamically casts the Decl to an `HLSLBufferDecl`, and then if it succeeds,
checks isCBuffer(). If true, the Decl is a cbuffer, otherwise it is a tbuffer. In either case,
we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
respectively). Otherwise, if the dynamic cast fails, we cast to `VarDecl`, and check 
to see if the ResourceAttr attribute is present. If it is, the `resource` flag can be 
set, and the corresponding resource class flag will be set. If the `VarDecl` dynamic cast

> fails, then there is some critical validation error, since any Decl with the register 

Reviewer 4: >If the `VarDecl` dynamic cast
fails, then there is some critical validation error
>
What is the fault taulerance strategy here? Are you intending to have a default case to handle this?

Author: I was intending to check if either decl is castable, and if not an llvm_unreachable is executed, "unknown decl type"

in the given UDT, we verify that the corresponding resource class flag has been set. These
are the register types `t`, `u`, `b`, and `s`. If the corresponding resource class flag is 
not set, `warn_hlsl_UDT_missing_resource_type_member` will be emitted.
If `c` is given, then we must check that `contains_numeric` is set, and if not, 
`warn_hlsl_UDT_missing_basic_type` is emitted.
Otherwise, if any other register type is given, 
then we emit `err_hlsl_unsupported_register_type_and_variable_type`.

All the warnings introduced in this spec were not emitted in legacy versions of the compiler.
The warnings, `warn_hlsl_register_type_c_not_in_global_scope`,
`warn_hlsl_deprecated_register_type_b`,
`warn_hlsl_deprecated_register_type_i`,
`warn_hlsl_UDT_missing_resource_type_member`, and
`warn_hlsl_UDT_missing_basic_type`, are all within a warning group known as 
`disallow-legacy-binding-rules`. However, only `warn_hlsl_UDT_missing_resource_type_member`
and  `warn_hlsl_UDT_missing_basic_type` are not treated as errors by default, 
the rest of the warnings are errors by default. If legacy behavior is desired, a 

> user can pass the `--Wno-disallow-legacy-binding-rules` flag to the compiler to silence the

Author: Yes, here, legacy means earlier versions of DXC.
And yes, I think it wouldn't hurt to make sure that flag works on a `clang-dxc` invocation. Because I can define warning groups in clang, I would expect the option can be easily carried forward.

> user can pass the `--Wno-disallow-legacy-binding-rules` flag to the compiler to silence the

Author: No further work is necessary, I've confirmed that this flag works on clang-dxc.exe.
// RUN: %clang_dxc -T lib_6_3 -Vd -fsyntax-only -Wno-disallow-legacy-binding-rules %s
runs fine on `clang\test\SemaHLSL\resource_binding_attr_error_silence_diags.hlsl`

diagnostics need to be emitted so that the user knows why the resource type is 
invalid, or why the register type isn't correct. This spec defines the behavior
of the compiler for whenever an invalid register type is given, or for when a valid
register type is used to bind a resource type that for whatever reason is 
incompatible with that register type.

## Motivation

There are several cases in DXC in which diagnostics that are related to
invalid register types are out of date or invalid. For example, in the case of:

`float b : register(u4);`

an error will be emitted recommending the use of the 'b, c, or i' register
type. However the 'i' register type is no longer in support, and the 'b' 
register type is only reserved for resource types that are constant buffers.
It is worth noting that there is an overloading of the register(...) keyword
using the 'c' register type to indicate an offset into a constant buffer for
numeric types only, as opposed to specifying a resource binding.
Additionally, it is possible the user is unaware that this variable won't 
actually be used as a resource, but the compiler doesn't communicate that 

> to the user. It would be great for any HLSL developer to immediately have
> answers to the following questions:
> For any valid variable type that needs to be bound as a resource,
> what is the set of valid register types for that variable,
> and what are all the possible diagnostics that can be emitted if the variable
> is given a register type that isn't in the set of valid register types?
> The design below aims to specify the answer to these questions.

Reviewer: Adding some bullet points to break this up a bit I think that the long sentence may have lost its way:

> It would be great for any HLSL developer to immediately have answers to the following questions:
> * For any valid variable type that needs to be bound as a resource,
>   * what is the set of valid register types for that variable, and 
>   * what are all the possible diagnostics that can be emitted if the variable is given a register type that isn't in the set of valid register types?

I don't think HLSL developers really care about what all the possible diagnostics are. Actually, the more I look at this the less difference I see between the last two bullet points above.

```suggestion
to the user. We should add diagnostics that inform the HLSL developer what register types are valid for that variable.
```

I'm not sure that much more is needed.

## Proposed solution

> Firstly, for future reference, the set of relevant diagnostics that would be emitted
> depending on how the variable type and register type interact should
> be named and defined. Below is a table that lists out all the proposed diagnostic
> names and diagnostic messages that are relevant in this spec:

Reviewer: ```suggestion
The following table lists all the proposed diagnostic
names and diagnostic messages that are relevant in this spec:

```

> | Diagnostic Name | Diagnostic message | Description |

Reviewer: There's a lot of different diagnostics here, and some of them seem extremely similar to me.

Also, I wonder about some of them...for example:

```
float f : register(x0); // error: invalid register type 'x' used; expected 't', 'u', 'b', or 's'

// attempt to fix:

float f : register(t0); // error: unsupported register binding 't' on variable of type 'float'
float f : register(u0); // error: unsupported register binding 'u' on variable of type 'float'
float f : register(b0); // error: unsupported register binding 'b' on variable of type 'float'
float f : register(s0); // error: unsupported register binding 's' on variable of type 'float'
```

So the first error gave me four expected register types, none of which work.

---

I think `err_hlsl_unsupported_register_type_and_resource_type`  should only be generated if one of these unsupported types is used on a UDT.  For all other cases we know _exactly_ what register type is valid and should specify it there.

Author: Thanks for the feedback.
I agree that `err_hlsl_unsupported_register_type_and_resource_type` should be left to the UDT case. For the Resource case, I can definitely change the diagnostic to specify the exact register type that is expected. But for Basic types I don't think the compiler should necessarily recommend using "c", because it is unknown what the user is trying to do. Recommending "c" is outside the scope of this diagnostic infrastructure because it has nothing to do with binding a resource. Updated the spec to reflect this.

| warn_hlsl_register_type_c_not_in_global_scope | "warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead" | Emitted if a basic type is bound with `c` within a `cbuffer` or `tbuffer` scope |
| warn_hlsl_deprecated_register_type_b | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." | Emitted if the register prefix `b` is used on a variable type without the resource class attribute, like a float type. |
| warn_hlsl_deprecated_register_type_i | "warning: deprecated legacy int constant register binding 'i' used." | Emitted if the register prefix `i` is used on a variable type without the resource class attribute, like a float type. |
| warn_hlsl_user_defined_type_missing_resource_type_member | "warning: variable of type '%0' bound to register type '%1' does not contain a matching '%select{srv\|uav\|cbv\|sampler}2' resource" | Emitted if a UDT is lacking any eligible member to bind using a specific register type, like if a UDT `Foo` was bound with `s` but had no sampler members.|

> | warn_hlsl_user_defined_type_missing_basic_type | "warning: register 'c' used on type with no contents to allocate in a constant buffer" | Emitted if a UDT has no numerical members, and is being bound using the `c` register type.|

Author: My issue with using `warn_hlsl_binding_type_mismatch` to replace `warn_hlsl_user_defined_type_missing_resource_type_member` and `warn_hlsl_user_defined_type_missing_basic_type`:
'c' is separate, because to my understanding, it isn't quite a register type, and it isn't binding a resource. So, the language used for 'c', "constant buffer data", should apply to register type 'b'. 'c' is for basic types that are to be placed in the $Globals constant buffer, which is a separate concept when compared to allocating a separate constant buffer, which is specified by 'b'. So I'm not quite sure 'c' is a "binding type".
I could change the message to include a missing type for each of the 5 register types (t u b s c), but I think the separation of concepts is significant enough to warrant a separate diagnostic in the "c" case.

Issue with `err_hlsl_binding_type_mismatch` replacing `err_hlsl_mismatching_register_type_and_resource_type` and `err_hlsl_unsupported_register_type_and_variable_type`:
There should be separate errors to indicate whether the register type is valid or invalid, and whether the register type is valid for a specific resource or not. Your suggested replacement diagnostic doesn't mention the register binding type that was used, which I think is helpful information. I could just rename the diagnostic to `err_hlsl_binding_type_mismatch` from `err_hlsl_mismatching_register_type_and_resource_type`, but I think the existing message is better, what do you think?

My issue with `err_hlsl_binding_type_invalid` replacing `err_hlsl_unsupported_register_type_and_variable_type`:
It is important the diagnostic lists "for variable type x" instead of just declaring the register type invalid. This is because the diagnostic gets emitted not only when the register type is invalid (like "y"), but also when a completely invalid variable type is attempted to be bound (like RayQuery). We emit the diagnostic in this case, so it could be that "t" is used with RayQuery, which is a valid register type. It's just that the variable type is invalid. We could take your suggestion, but that would require adding a new diagnostic for cases where the variable type is all-around invalid. Thoughts?

I can accept the rename to `err_hlsl_duplicate_register_annotation`
I can accept the rename to `warn_hlsl_register_type_c_packoffset`
I can accept both rewording suggestions.

> | warn_hlsl_user_defined_type_missing_basic_type | "warning: register 'c' used on type with no contents to allocate in a constant buffer" | Emitted if a UDT has no numerical members, and is being bound using the `c` register type.|

Reviewer 5: > My issue with using `warn_hlsl_binding_type_mismatch` to replace `warn_hlsl_user_defined_type_missing_resource_type_member` and `warn_hlsl_user_defined_type_missing_basic_type`: 'c' is separate, because to my understanding, it isn't quite a register type, and it isn't binding a resource. So, the language used for 'c', "constant buffer data", should apply to register type 'b'. 'c' is for basic types that are to be placed in the $Globals constant buffer, which is a separate concept when compared to allocating a separate constant buffer, which is specified by 'b'. So I'm not quite sure 'c' is a "binding type". I could change the message to include a missing type for each of the 5 register types (t u b s c), but I think the separation of concepts is significant enough to warrant a separate diagnostic in the "c" case.

The way I see it, the 'b' binding type is for a constant buffer. The 'c' binding type is for "constant buffer data (in the $Globals constant buffer)". Given that 'b' bindings are *never* valid on a VarDecl this separation seemed clean to me, and I think misuses of the 'b' type are the actual outlier here (which is covered by the `warn_hlsl_deprecated_register_type_b` diagnostic).

> Issue with `err_hlsl_binding_type_mismatch` replacing `err_hlsl_mismatching_register_type_and_resource_type` and `err_hlsl_unsupported_register_type_and_variable_type`: There should be separate errors to indicate whether the register type is valid or invalid, and whether the register type is valid for a specific resource or not. Your suggested replacement diagnostic doesn't mention the register binding type that was used, which I think is helpful information. I could just rename the diagnostic to `err_hlsl_binding_type_mismatch` from `err_hlsl_mismatching_register_type_and_resource_type`, but I think the existing message is better, what do you think?

My thoughts were that this error is for "the binding type is valid but it's applied to something that doesn't make sense". I found the proposed set of messages for this case to be kind of confusing and inconsistent, so the suggestion here is to not bother trying to classify what exactly the binding is being applied to, but just say that it doesn't make sense and explain what it *does* apply to.

> My issue with `err_hlsl_binding_type_invalid` replacing `err_hlsl_unsupported_register_type_and_variable_type`: It is important the diagnostic lists "for variable type x" instead of just declaring the register type invalid. This is because the diagnostic gets emitted not only when the register type is invalid (like "y"), but also when a completely invalid variable type is attempted to be bound (like RayQuery). We emit the diagnostic in this case, so it could be that "t" is used with RayQuery, which is a valid register type. It's just that the variable type is invalid. We could take your suggestion, but that would require adding a new diagnostic for cases where the variable type is all-around invalid. Thoughts?

If 't' is used with RayQuery we should just emit `err_hlsl_binding_type_mismatch` - RayQuery is not an SRV, end of story. Having a simple "This binding type is never valid" is simple and clear, so it makes sense to be its own thing.

> I can accept the rename to `err_hlsl_duplicate_register_annotation` I can accept the rename to `warn_hlsl_register_type_c_packoffset` I can accept both rewording suggestions.

> | warn_hlsl_user_defined_type_missing_basic_type | "warning: register 'c' used on type with no contents to allocate in a constant buffer" | Emitted if a UDT has no numerical members, and is being bound using the `c` register type.|

Reviewer 5: > > My issue with using `warn_hlsl_binding_type_mismatch` to replace `warn_hlsl_user_defined_type_missing_resource_type_member` and `warn_hlsl_user_defined_type_missing_basic_type`: 'c' is separate, because to my understanding, it isn't quite a register type, and it isn't binding a resource. So, the language used for 'c', "constant buffer data", should apply to register type 'b'. 'c' is for basic types that are to be placed in the $Globals constant buffer, which is a separate concept when compared to allocating a separate constant buffer, which is specified by 'b'. So I'm not quite sure 'c' is a "binding type". I could change the message to include a missing type for each of the 5 register types (t u b s c), but I think the separation of concepts is significant enough to warrant a separate diagnostic in the "c" case.
> 
> The way I see it, the 'b' binding type is for a constant buffer. The 'c' binding type is for "constant buffer data (in the $Globals constant buffer)". Given that 'b' bindings are _never_ valid on a VarDecl this separation seemed clean to me, and I think misuses of the 'b' type are the actual outlier here (which is covered by the `warn_hlsl_deprecated_register_type_b` diagnostic).

Ah, this isn't quite correct though, the 'b' binding is on a `VarDecl` here:

```hlsl
struct cbdata {
  float4 position;
};
ConstantBuffer<cbdata> cb1 : register(b);
```

even so, we have the special-cased separate diagnostic for when 'b' isn't correct so I don't think it changes much in practice.

> | warn_hlsl_user_defined_type_missing_basic_type | "warning: register 'c' used on type with no contents to allocate in a constant buffer" | Emitted if a UDT has no numerical members, and is being bound using the `c` register type.|

Author: On further thought I decided to go ahead and take your suggestions, it makes diagnostic interpretation easier.

Note that there are instances where multiple `register` statements can apply to a single 
decl, and this is only invalid if any of the register annotations have the same register types.
If this invalid case happens, `err_hlsl_conflicting_register_annotations` will be emitted. 
There is a separate case where two decls in separate locations in the translation unit have 
overlapping register numbers and register types, causing a conflict. This type of conflict 
cannot be detected at this stage of compilation, because parsing is not yet complete. 
Detecting this conflict is out of scope for this diagnostic infrastructure, but will be 
caught later by the register allocation algorithm. 
In `dxc`, `\lib\HLSL\DxilCondenseResources.cpp` has a class called
`DxilResourceRegisterAllocator` with a member `AllocateRegisters` that is responsible for
allocating registers and validating that there aren't any conflicts or overlaps. As for 
`clang-dxc`, there is not yet any register allocation validation, but when resources are
finalized, allocation validation must be implemented, and will likely use the same algorithm 

> used in DXC.

Reviewer 2: See my comment on that proposal regarding the major breaking change this proposal implies to prior HLSL behavior:
https://github.com/llvm/llvm-project/issues/58051#issuecomment-2239952237

The design below aims to specify the answer to these questions.

## Proposed solution
The following table lists all the proposed diagnostic
names and diagnostic messages that are relevant in this spec:

| Diagnostic Name | Diagnostic message | Description |
|-|-|-|
| err_hlsl_mismatching_register_type_and_variable_type | "error: unsupported resource register binding '%select{t\|u\|b\|s}0' on variable of type '%1'"| Emitted if a variable type that isn't accompanied with a resource class attribute is bound with a standard register type (`t`, `u`, `b`, or `s`). |
| err_hlsl_unsupported_register_type_and_variable_type | "error: register binding type '%0' not supported for variable of type '%1'" | Emitted if a variable type is bound using an unsupported register type, like binding a float with the 'x' register type. |
| err_hlsl_mismatching_register_type_and_resource_type | "error: %select{srv\|uav\|cbv\|sampler}2 type '%0' requires register type '%select{t\|u\|b\|s}2', but register type '%1' was used." | Emitted if a known resource type is bound using a standard but mismatching register type, e.g., RWBuffer<int> getting bound with 's'.|

> | err_hlsl_unsupported_register_type_and_resource_type | "error: invalid register type '%0' used; expected 't', 'u', 'b', or 's'"| Emitted if an unsupported register type is used to bind a resource, like 'y' being used with RWBuffer<int>.|

Reviewer: Is this up to date with the latest design?

Author: The actual implementation needs some updating given the new handling for when invalid registers are detected. Otherwise, I do believe the spec is self-consistent now.

`tbuffer {...}` scope, the register number will be ignored. In this case, 
`warn_hlsl_register_type_c_not_in_global_scope` will be emitted, and will be treated as an error by default.
All other legal register types applied to such a type will emit `err_hlsl_mismatching_register_type_and_variable_type`, 
except for 'b' or 'i'. If the register type is 'b' or 'i', then `warn_hlsl_deprecated_register_type_b` or 
`warn_hlsl_deprecated_register_type_i` will be emitted respectively, and treated as errors by default. 
The compiler will function as if `c` was passed. 
Invalid register types will cause `err_hlsl_unsupported_register_type_and_variable_type` to be emitted.

Below are some examples:

| Code | Diagnostic |
|-|-|
| `float f : register(t0)` | "error: unsupported resource register binding 't' on variable of type 'float'" |
| `float f : register(c0)` | Valid, no errors
| `float f : register(b0)` | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." |
| `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used." |

> | `float f : register(x0)` | "error: invalid register type 'x' used; expected 't', 'u', 'b', or 's'" |

Reviewer: Does this need to be updated as part of resolving https://github.com/microsoft/hlsl-specs/pull/230/files#r1681673518?

types of members in the UDT are defined. Types of interest may be contained in arrays
of arbitrary dimension, so UDT analysis also requires drilling into arrays to determine
if the underlying element type is numeric, or any resource class type of interest.
If the decl is not a UDT, then the `other` flag is set.

Below is some pseudocode to describe the flag-setting process:

```
if the decl is groupshared:
  set other flag

if the Decl is not inside a cbuffer or tbuffer:
  if the Decl type is a numeric type:
    set DefaultGlobals

if the Decl is an HLSLBufferDecl:
  set resource
  if cbuffer: set cbv
  else: set srv
else if the Decl is a VarDecl:
  if resource attribute: 
    get resource attribute, set resource and corresponding resource class flag
  else:
    if Decl is vector, matrix, or otherwise numeric type:
      set basic flag
    else if Decl is struct or class type (if it is a udt):
      set udt flag
      recurse through members, set resource class flags that are found
      set ContainsNumeric if the UDT contains a numeric member
    else:
      set other flag
else:
  raise error: (unknown decl type)
```

### Diagnostic emisison

> Now enough information has been gathered to determine the right diagnostics to emit, if any.

Reviewer: I don't think I'm going to be able to read this all carefully enough to verify that it is correct. I'm basing most of my review on the tables of diagnostics and the examples shown above. I suspect that the full details of _how_ these diagnostics are determined would be best described in code.

Reviewer: (resolving: no action necessarily required)

Additionally, it is possible the user is unaware that this variable won't 
actually be used as a resource, but the compiler doesn't communicate that 
to the user. We should make it clear in this document which variables are 
compatible with which register types.

## Proposed Solution

The resource binding attribute will be attached to any declaration object (Decl)
that has the `: register(...)` annotation. In Sema, this attribute has a function to
validate its correctness, called `handleResourceBindingAttr`, within 
`clang\lib\Sema\SemaHLSL.cpp`. The diagnostic infrastructure will be implemented 
within this validation function to analyze the declaration that the annotation
is applied to, and validate that the register type used within the annotation is 
compatible with the Decl type. There are other details that need validation, such
as whether there are any duplicate register annotations with the same register 
type that have been applied to the Decl. Even the context in which the Decl
appears (inside or outside a cbuffer or tbuffer) influences the legality of 
certain register types. All of this analysis and validation will be executed

> inside a new function, `DiagnoseHLSLResourceRegType`.

Author: Makes sense, updated.

There are 4 register types that are expected to be used to bind common resources:

| Resource Class | Register Type |
|-|-|
| SRV | t |
| UAV | u |
| CBuffer | b |
| Sampler | s |

`DiagnoseHLSLResourceRegType` will validate that resources are bound using the 
expected register type. There are other register types that may legally appear in 

> the register annotation, `c` and `i`. `DiagnoseHLSLResourceRegType` will be 

Author: Took out the word `legally`, what I mean is that the appearance of `i` will cause a unique behavior in the compiler that is worth describing.

responsible for determining if these register types are used correctly in certain
legacy contexts, or whether such uses are invalid. `DiagnoseHLSLResourceRegType` 
will also be responsible for emitting a diagnostic if any other invalid register
type is detected. This infrastructure will prioritize emitting diagnostics that
explain why the variable type isn't suitable for the register type, rather than 

> why the register type isn't suitable for the variable type.

Reviewer 2: > This infrastructure will prioritize emitting diagnostics that
> explain why the variable type isn't suitable for the register type, rather than 
> why the register type isn't suitable for the variable type.

I believe this comment is being made based on some feedback last week on some diagnostic wording, but I'm not sure from this high-level context what to make of the statement - why is it here, what does it mean exactly?

Obviously, we will want to emit a diagnostic that indicates a mismatch between these things.  This statement makes it sound like the register type in the annotation should be the primary thing and we expect someone needs to adjust the variable type to match expectations set by the register binding, when intuitively it's probably the other way around.

Author: Yeah, this seems like it's more of an implicit philosophy on diagnostic wording that doesn't need to be called out in the spec, I removed this idea.

If `DiagnoseHLSLResourceRegType` finds any critical errors, the attribute,
`handleResourceBindingAttr`, won't be added to the Decl, and compilation will
fail. However, `handleResourceBindingAttr` may emit some warnings and allow

> the attribute to be attached.

Author: It's a division of syntactic and semantic validation responsibilities. I'll add that detail to the spec.

> the attribute to be attached.

Reviewer 2: I wouldn't be too concerned with this if you plan on taking the advice in my overall feedback and leave these sorts of details up to the detailed design section instead.

> the attribute to be attached.

Reviewer 2: To be clear, it still makes sense to categorize diagnostic scenarios in this section based on the information required and thus when/where they will be performed.

`float b : register(u4);`
an error will be emitted recommending the use of the 'b, c, or i' register
type. However the 'b' and 'i' register types are no longer in support. It
is worth noting that there is an overloading of the register(...) keyword

> that applies to constant buffer variables using the 'c' register type, as

Reviewer 2: Not "constant buffer variables", since that's still confusing (`ConstantBuffer<Foo> myCB` looks like a constant buffer variable).  How about:
```suggestion
using the 'c' register type to indicate an offset into a constant buffer, as
```

> opposed to resources. 

Reviewer 2: ```suggestion
opposed to specifying a resource binding.
```

names and diagnostic messages that are relevant in this spec:

| Diagnostic Name | Diagnostic message | Description |
|-|-|-|
| err_hlsl_mismatching_register_type_and_variable_type | "error: unsupported resource register binding '%select{t|u|b|s}1' on variable of type '%0'"|  |
| err_hlsl_unsupported_register_prefix | "error: invalid resource class specifier '%0' used; expected 't', 'u', 'b', or 's'"| |
| err_hlsl_mismatching_register_type_and_resource_type | "error: %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used." | |
| err_hlsl_mismatching_unsupported_register_type_and_variable_type | "error: register binding type '%1' not supported for variable of type '%0'" | |
| warn_hlsl_register_type_c_not_in_global_scope | "warning: register binding 'c' ignored inside cbuffer/tbuffer declarations; use pack_offset instead" | This is emitted when a basic type is bound with `c` within a `cbuffer` or `tbuffer` scope |
| warn_hlsl_deprecated_register_type_b | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding." | |

> | warn_hlsl_deprecated_register_type_i | "warning: deprecated legacy int constant register binding 'i' used." | |

Author: There don't seem to be any behavioral differences between the two register types in DXC, so I'm willing to say yes, `b` does behave as if `c` was specified.

Firstly, the most common case is when a resource type that is not a 
user-defined type is bound as a resource. Any resource will have an 
associated ResourceAttr attribute, from which we can determine the 

> resource class The resource class data is sufficient to determine 

Reviewer 2: ```suggestion
resource class, which is sufficient to determine 

```

the expected register type, and any other register type will result 
in an error. The recommended register type will be suggested in the
diagnostic. The table below specifies what register type will be 
expected for any resource type that falls under the specified 
resource class. 

| Resource Class | Register Type | Diagnostic |
|-|-|-|
| Sampler | s | "%select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used." |
| SRV | t | "%select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used." |
| UAV | u | "%select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used." |

> | CBuffer | b | "%select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used." |

Reviewer 2: The diagnostic string in the table isn't useful now, since it's identical for each row.  It would be good to provide the diagnostic string in one place (before the table).  Then this is just a table of Resource Class to Register Type.

> | CBuffer | b | "%select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used." |

Reviewer: You may want to reconsider formatting this as a table, since I think the Diagnostic column is identical for all of the rows now? It looks like they are, but listing it four times implies that there's some subtle difference I'm missing.

If the given candidate resource is a user-defined type (UDT), then further
analysis is necessary. The first step is to gather all register declarations
that are being applied to this variable declaration, and collect the register types
that are being specified. The UDT must contain at least one valid resource
that can be bound to the provided register type(s). If not, an error must be

> emitted stating that "No resource contained in struct '%0' can be bound

Author: We did, forgot to add that default error note, even though it was present in the examples. 
Made a note of your second point.

to register type '%1'". There are no issues if a UDT contains more resources
than there are register binding statements, the resources will be bound to
the next available space automatically, and so compilation can succeed.
Below are some examples of different UDT's and the diagnostics that

> would be emitted when the resources contained within them are bound:

Reviewer 2: ```suggestion
would be emitted when applying resource bindings to the variable:
```

| err_hlsl_mismatching_register_type_and_variable_type | "error: unsupported resource register binding '%select{t\|u\|b\|s}0' on variable of type '%1'"| Emitted if a variable type that isn't accompanied with a resource class attribute is bound with a standard register type (`t`, `u`, `b`, or `s`). |
| err_hlsl_unsupported_register_type_and_variable_type | "error: register binding type '%0' not supported for variable of type '%1'" | Emitted if a variable type is bound using an unsupported register type, like binding a float with the 'x' register type. |
| err_hlsl_mismatching_register_type_and_resource_type | "error: %select{srv\|uav\|cbv\|sampler}2 type '%0' requires register type '%select{t\|u\|b\|s}2', but register type '%1' was used." | Emitted if a known resource type is bound using a standard but mismatching register type, e.g., RWBuffer<int> getting bound with 's'.|

> | err_hlsl_unsupported_register_type_and_resource_type | "error: invalid register type '%0' used; expected 't', 'u', 'b', or 's'"| Emitted if an unsupported register type is used to bind a UDT.|

Reviewer: "err_hlsl_unsupported_register_type_and_resource_type" - do these diagnostic names appear anywhere user-visible?  I don't think the name really matches what it is now.  I'd expect "user defined type" to be in the name somewhere, for example.  Not sure "resource type" belongs at all.

Author: They do not, but because UDTs are now the only use, it makes sense to update the name.

| UAV | u | "Object type '%0' with resource class "UAV" expects register type 'u', but register type '%1' was given." |
| CBuffer | b |  "Object type '%0' with resource class "CBuffer" expects register type 'b', but register type '%1' was given." |

There are certain exceptions to the table above. While we would like to
reserve a register type for each resource class, and create a one-to-one
mapping, there is legacy behavior that must be preserved in two specific
cases. A Sampler resource class may use the 't' register type, and an
SRV resource class may use the 's' register type. However, the compiler
will not actively use such resources. This legacy behavior was permitted
in prior versions of the compiler, but will now emit a warning that will
be treated as an error by default. So, there will be a special exception
for Samplers and SRVs, which may accept 't' and 's' respectively if the

> warning (--Wno-disallow-legacy-binding-rules) is disabled.

Reviewer: We've discussed this, but I figured I'd get a note on here that I'm unconvinced we need to preserve this behavior.  FXC didn't support this, and the behavior in DXC seems more like a bug than a legacy feature.

Author: Yes, we decided that this behavior should no longer be supported, so this paragraph has now been removed from the spec.

"intangible". In this case, we are dealing with types that cannot be intangible.
Types that can be immediately determined to not be intangible (that is, types that
cannot be a resource type or cannot contain a resource type) are types like booleans,
int, float, float4, etc. If the variable type is among any numerics or a type that
cannot be an intangible type, there is only one exception where the register keyword
can legally be applied to the variable. The only valid register type for such a 
numeric variable type is 'c'. The register keyword in this statement:
`float f : register(c0)`
isn't binding a resource, rather it is specifying a constant register binding offset

> within the $Globals cbuffer, which is legacy behavior from DX9. All other register types

Reviewer 2: Another note here:
The context of the declaration where this is allowed is restricted to constants that would be placed into `$Globals` because they are outside of any `cbuffer {}` or `tbuffer {}` declaration.  In DXC and FXC, when the register binding is used in these other contexts it is silently ignored.  We should decide whether we want a DefaultError warning or hard error in that case.

Author: I think default error warning for now, and we can make it more strict and make it an error in 202x.

If 't', 'u', or 's' are given, then this diagnostic will be emitted:
`error: '%0' is an invalid resource type for register type '%1'`. If any other register type
is seen, this error will be emitted:
`error: register binding type '%1' not supported for variable of type '%0'`<br><br>

Finally, in the case that `udt` is set, we first check `default_globals`. If it is set,
then we can permit the 'c' register type. After this point, `default_globals` doesn't
need to be set. For every register type that is used to bind the resources contained in
the given UDT, we verify that the corresponding resource class flag has been set. If the
corresponding resource class flag is not set, this error will be emitted:
`error: variable of type '%0' bound to register type '%1' does not contain a matching '%select{SRV|UAV|CBV|Sampler}2' resource`
Otherwise, if any other register type is given, then we emit this error:
`error: register binding type '%0' not supported for variable of type '%1'`<br><br>

Legacy behavior can be allowed with the --Wno-disallow-legacy-binding-rules
flag. When this flag is active, and legacy behavior is present, a warning will 
be emitted instead of an error.<br><br>

> Below are two flowcharts that describe the process that is used to determine

Reviewer: I'm sorry, I'm really not a fan of the flowcharts.  I find them hard to read in the previews: they're so zoomed out I can't read the text, and if I zoom in then I have to use the clunky navigation things that github adds to them to look around.  And they're essentially write-only in the raw md view.

Reviewer: ![image](https://github.com/microsoft/hlsl-specs/assets/8118402/fc98d803-ddc7-42f9-b724-b70c3f9e09b0)

Author: I do think that they're useful since it's difficult to describe this entire process with just words. Lots of conditions, and it's not always clear where exactly if's `end` unless you're extremely verbose. But I'll remove them for now and see if the description is acceptable.

};

Eg9 e9;
// "error: location annotations cannot be specified on members"

template<typename R>
struct Eg10 {
    R b;
};

Eg10<Texture2D> t : register(u0);

> // "warning: variable of type 'Eg10' bound to register type 'u' does not contain a matching 'UAV' resource"

Reviewer: Does the requirement to look inside template expansions change anything about the plan on implementing this in handleHLSLResourceBindingAttr?

Author: `RequireCompleteType` needs to be called in template expansion cases, but there doesnt seem to be any fundamental reason why this specific location is insufficient for implementing this infrastructure.

If any other register type is seen, `err_hlsl_unsupported_register_type_and_variable_type`
will be emitted instead.

> Finally, in the case that `udt` is set, we fmay have multiple register annotations

Reviewer: ```suggestion
Finally, in the case that `udt` is set, we may have multiple register annotations

```

      emit err_hlsl_unsupported_register_type_and_variable_type  
```

## Behavioral Differences

> This infrastructure will introduce some behavioral differences between `clang-dxc` and `dxc`.

Reviewer: I don't think you're intending to imply that there will be further differences between clang-dxc and clang? If we're describing the behavior in clang we should probably refer to it as clang. If we're referring to any behavior that's specific to when running through the clang-dxc interface then we should call that out then.

Reviewer 4: I'm not sure I come to the same conclusion @damyanp. The line wants to distinguish behavioral differences between `clang-dxc` and  dxc. I don't see any mention of `clang`

Reviewer: My point here is that clang-dxc is just the driver for running clang with the various flags set up so that the command-line looks like dxc.  The thing we're actually building (and what this document is describing) is how clang behaves.

Author: I agree, changed to `clang`
